### ISSUE TITLE
Code transformed incorrectly

### ISSUE THREAD

**Issue by z4o4z** [rollup/rollup#4960]
### Rollup Version

3.21.0

### Operating System (or Browser)

MacOS Version 13.3.1 (22E261)

### Link To Reproduction

[short example](https://rollupjs.org/repl/?version=3.21.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQSUyMiUyMiUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJjb2RlJTIyJTNBJTIyY29uc3QlMjBtb2RpZnklMjAlM0QlMjAoJTdCJTIwY2FsbGJhY2tzJTIwJTdEKSUyMCUzRCUzRSUyMCU3QiU1Q24lMjAlMjBjYWxsYmFja3Mua2V5QmluZGluZ0ZuJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlN0QlNUNuJTIwJTIwY2FsbGJhY2tzLmhhbmRsZVJldHVybiUyMCUzRCUyMCgpJTIwJTNEJTNFJTIwJTdCJTdEJTVDbiUyMCUyMGNhbGxiYWNrcy5oYW5kbGVLZXlDb21tYW5kJTIwJTNEJTIwKCklMjAlM0QlM0UlMjAlN0IlN0QlNUNuJTdEJTVDbiU1Q24lNUNud2luZG93LmluZGV4JTIwJTNEJTIwKGZ1bmN0aW9uJTIwKCklMjAlN0IlNUNuJTIwJTIwdmFyJTIwY29uZmlnJTIwJTNEJTIwYXJndW1lbnRzLmxlbmd0aCUyMCUzRSUyMDAlMjAlMjYlMjYlMjBhcmd1bWVudHMlNUIwJTVEJTIwISUzRCUzRCUyMHVuZGVmaW5lZCUyMCUzRiUyMGFyZ3VtZW50cyU1QjAlNUQlMjAlM0ElMjAlN0IlN0QlM0IlNUNuJTIwJTIwdmFyJTIwY2FsbGJhY2tzJTIwJTNEJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGtleUJpbmRpbmdGbiUzQSUyMHVuZGVmaW5lZCUyQyU1Q24lMjAlMjAlMjAlMjBoYW5kbGVLZXlDb21tYW5kJTNBJTIwdW5kZWZpbmVkJTJDJTVDbiUyMCUyMCUyMCUyMGhhbmRsZVJldHVybiUzQSUyMHVuZGVmaW5lZCUyQyU1Q24lMjAlMjAlMjAlMjBoYW5kbGVQYXN0ZWRUZXh0JTNBJTIwdW5kZWZpbmVkJTJDJTVDbiUyMCUyMCUyMCUyMG9uQ2hhbmdlJTNBJTIwdW5kZWZpbmVkJTVDbiUyMCUyMCU3RCUzQiU1Q24lMjAlNUNuJTIwJTIwJTVDbiUyMCUyRiUyRiUyMC4uLi4lNUNuJTIwJTIwdmFyJTIwcHJvcHMlMjAlM0QlMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTJGJTJGJTIwLi4uJTVDbiUyMCUyMCUyMCUyMGNhbGxiYWNrcyUyQyU1Q24lMjAlMjAlN0QlM0IlNUNuJTIwJTIwJTVDbiU1Q24lMjAlMjByZXR1cm4lMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTJGJTJGJTIwLi4uLiU1Q24lMjAlMjAlMjAlMjBtZXRob2QlM0ElMjAoKSUyMCUzRCUzRSUyMG1vZGlmeSgpJTJDJTVDbiUyMCUyMCUyMCUyMGtleUJpbmRpbmdGbiUzQSUyMGZ1bmN0aW9uJTIwa2V5QmluZGluZ0ZuKGtleWJvYXJkRXZlbnQpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMGlmJTIwKGNhbGxiYWNrcy5rZXlCaW5kaW5nRm4pJTIwcmV0dXJuJTIwY2FsbGJhY2tzLmtleUJpbmRpbmdGbihrZXlib2FyZEV2ZW50KSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjByZXR1cm4lMjB1bmRlZmluZWQlM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMGhhbmRsZVJldHVybiUzQSUyMGZ1bmN0aW9uJTIwaGFuZGxlUmV0dXJuKGtleWJvYXJkRXZlbnQpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMGlmJTIwKGNhbGxiYWNrcy5oYW5kbGVSZXR1cm4pJTIwcmV0dXJuJTIwY2FsbGJhY2tzLmhhbmRsZVJldHVybihrZXlib2FyZEV2ZW50KSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjByZXR1cm4lMjB1bmRlZmluZWQlM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTJDJTVDbiUyMCUyMCUyMCUyMGhhbmRsZVBhc3RlZFRleHQlM0ElMjBmdW5jdGlvbiUyMGhhbmRsZVBhc3RlZFRleHQoa2V5Ym9hcmRFdmVudCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwaWYlMjAoY2FsbGJhY2tzLmhhbmRsZVBhc3RlZFRleHQpJTIwcmV0dXJuJTIwY2FsbGJhY2tzLmhhbmRsZVBhc3RlZFRleHQoa2V5Ym9hcmRFdmVudCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwcmV0dXJuJTIwdW5kZWZpbmVkJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCUyQyU1Q24lMjAlMjAlMjAlMjBvbkNoYW5nZSUzQSUyMGZ1bmN0aW9uJTIwb25DaGFuZ2UoZWRpdG9yU3RhdGUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMGlmJTIwKGNhbGxiYWNrcy5vbkNoYW5nZSklMjByZXR1cm4lMjBjYWxsYmFja3Mub25DaGFuZ2UoZWRpdG9yU3RhdGUpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMGVkaXRvclN0YXRlJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lMjAlMjAlN0QlM0IlNUNuJTdEKSUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTdEJTdEJTdE)

I use library that has pretty weird code, it passes `callbacks` object into react components and other functions, and these components/functions mutate callbacks object. Other functions uses `callbacks` object to perform some logic, but rollup drops `callbacks` from them. It introduces really weird issues in our product. The example link is not perfect one (it doesn't use the props object as an argument), but it shows exactly what happens in my production bundle . If remove initial nullable values from the `callbacks` object, all works fine.

### Expected Behaviour

```js
 keyBindingFn: function keyBindingFn(keyboardEvent) {
      if (callbacks.keyBindingFn) return callbacks.keyBindingFn(keyboardEvent);
      return undefined;
    },
    handleReturn: function handleReturn(keyboardEvent) {
      if (callbacks.handleReturn) return callbacks.handleReturn(keyboardEvent);
      return undefined;
    },
    handlePastedText: function handlePastedText(keyboardEvent) {
      if (callbacks.handlePastedText) return callbacks.handlePastedText(keyboardEvent);
      return undefined;
    },
```

### Actual Behaviour

```js
  keyBindingFn: function keyBindingFn(keyboardEvent) {
      return undefined;
    },
    handleReturn: function handleReturn(keyboardEvent) {
      return undefined;
    },
    handlePastedText: function handlePastedText(keyboardEvent) {
      return undefined;
    },
```

**Comment by dnalborczyk** [comment]
just glancing over the repro, the outcome appears to be correct.

callbacks.keyBindingFn is `undefined`, therefore `if (callbacks.keyBindingFn) return callbacks.keyBindingFn(keyboardEvent);` is being removed.

**Comment by lukastaegert** [comment]
Yes, and since you never call `modify` with a parameter, it is impossible for `callbacks` to be changed.

**Comment by z4o4z** [comment]
here's the library bundled code, as you can see `callbaks` are used in the `DecoratedMentionSuggestionsComponent` as props and later in the return object methods
![image](https://user-images.githubusercontent.com/9130301/234823615-ab24294a-a930-4941-a867-1343c33e0f51.png). 

But in the application bundle it looks like this 
![image](https://user-images.githubusercontent.com/9130301/234824208-cd050160-ed5e-49fb-86e0-e9c82bdc7cb5.png)

I'm using vite to bundle my app. But minification is disabled in my tests, so I assume it's a rollup issue. Probably I'm wrong

**Comment by lukastaegert** [comment]
In your original code, where is the place that updates `callbacks`?

**Comment by z4o4z** [comment]
have no idea, i'm using `draftjs` editor with plugins and one of these plugins mutate callbacks. In runtime, i see that callbacks are mutated, functions are there

**Comment by dnalborczyk** [comment]
@lukastaegert 
from the first screenshot, `callbacks` are attached to an object `mentionSearchProps`

<img width="195" alt="image" src="https://user-images.githubusercontent.com/2903325/234880842-ca55f02d-22cb-4738-8623-a8560b4aead0.png">

which is being passed on to `React.createElement`:

<img width="671" alt="image" src="https://user-images.githubusercontent.com/2903325/234881004-f43c2c12-1581-403f-82ee-0a76c079cb68.png">

which could and probably does get mutated afterwards somehwere.

**Comment by z4o4z** [comment]
> @lukastaegert from the first screenshot, `callbacks` are attached to an object `mentionSearchProps`
> 
> <img alt="image" width="195" src="https://user-images.githubusercontent.com/2903325/234880842-ca55f02d-22cb-4738-8623-a8560b4aead0.png">
> 
> which is being passed on to `React.createElement`, with the result of `_extends`:
> 
> <img alt="image" width="671" src="https://user-images.githubusercontent.com/2903325/234881004-f43c2c12-1581-403f-82ee-0a76c079cb68.png">
> 
> which might either get merged/extended by the passed in props `(_extends(...))` or could get mutated afterwards somewhere.

exactly, i can add a proxy to find a place where callbacks are mutated, if we really need it

**Comment by lukastaegert** [comment]
Ah, thanks, I overlooked that. I fear we will need a better reproduction to dig into this. I don’t suppose you have a repository to share? Would also be interesting if Rollup 3.18.0 already had this issue or if 3.19.0 introduced it. It looks like _extends does not deoptimize its argument.

**Comment by lukastaegert** [comment]
What is the code of the `_extends` function in this example?

**Comment by z4o4z** [comment]
`_extends` is pretty simple, just a polyfill for `Object.assign`:

```js
function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}
```

**Comment by dnalborczyk** [comment]
could you create a repository with a minimal repro @z4o4z ? otherwise it's nearly impossible to tell what's going on.

**Comment by lukastaegert** [comment]
That may not be needed, I think I figured it out. It is the use of `arguments` in `_extends` that breaks it, see [this example](https://rollupjs.org/repl/?version=3.21.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwbXV0YXRlKCklMjAlN0IlNUNuJTIwJTIwYXJndW1lbnRzJTVCMCU1RC54JTIwJTNEJTIwdHJ1ZSUzQiU1Q24lN0QlNUNuJTVDbnZhciUyMG9iaiUyMCUzRCUyMCU3QiU1Q24lMjAlMjB4JTNBJTIwZmFsc2UlNUNuJTdEJTNCJTVDbiU1Q25tdXRhdGUob2JqKSUzQiU1Q24lNUNuaWYlMjAob2JqLngpJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCdtdXRhdGVkIScpJTNCJTVDbiU3RCUyMGVsc2UlMjAlN0IlNUNuJTIwJTIwY29uc29sZS5sb2coJ0ZBSUxFRCcpJTNCJTVDbiU3RCUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE).

This was not a problem prior to version 3.19.0 because here, call arguments were deoptimized by default, see [here](https://rollupjs.org/repl/?version=3.18.0&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwbXV0YXRlKCklMjAlN0IlNUNuJTIwJTIwYXJndW1lbnRzJTVCMCU1RC54JTIwJTNEJTIwdHJ1ZSUzQiU1Q24lN0QlNUNuJTVDbnZhciUyMG9iaiUyMCUzRCUyMCU3QiU1Q24lMjAlMjB4JTNBJTIwZmFsc2UlNUNuJTdEJTNCJTVDbiU1Q25tdXRhdGUob2JqKSUzQiU1Q24lNUNuaWYlMjAob2JqLngpJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCdtdXRhdGVkIScpJTNCJTVDbiU3RCUyMGVsc2UlMjAlN0IlNUNuJTIwJTIwY29uc29sZS5sb2coJ0ZBSUxFRCcpJTNCJTVDbiU3RCUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE).

I will see what I can do.

**Comment by lukastaegert** [comment]
Potential fix at #4965
